sycl/include/sycl/detail/kernel_name_based_cache.hpp

```diff
@@ -0,0 +1,36 @@
+//==--------------------- kernel_name_based_cache.hpp ----------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include <sycl/detail/export.hpp>
+
+#include <memory>
+
+namespace sycl {
+inline namespace _V1 {
+namespace detail {
+
+struct KernelNameBasedCacheT;
+struct __SYCL_EXPORT KernelNameBasedCacheDeleterT {
+  void operator()(KernelNameBasedCacheT *Ptr);
+};
+using KernelNameBasedCachePtrT =
+    std::unique_ptr<KernelNameBasedCacheT, KernelNameBasedCacheDeleterT>;
+__SYCL_EXPORT KernelNameBasedCachePtrT createKernelNameBasedCache();
+
+// Retrieves a cache pointer unique to a kernel name type that can be used to
+// avoid kernel name based lookup in the runtime.
+template <typename KernelName>
+KernelNameBasedCacheT *getKernelNameBasedCache() {
+  static KernelNameBasedCachePtrT Instance{createKernelNameBasedCache()};
```

Comment:
Can the `createKernelNameBasedCache` return a raw pointer? We can make the `using KernelNameBasedCachePtrT =
    *KernelNameBasedCacheT` and control the life time of the `KernelNameBasedCacheT` objects inside runtime, somewhere in GlobalHandler.
Whith the current approach the lifetime of the `KernelNameBasedCacheT` objects is controlled by the client of SYCL RT. If the destructor of the `KernelNameBasedCacheT` is not trivial and depends on SYCL RT it might be a problem because the SYCL RT might be already unloaded when the destructor of the `KernelNameBasedCacheT` is called.

Reply:
Yeah, that's a good point. As of right now it wouldn't really matter, but we definitely should future-proof it. I'll make the change.

Reply:
My main point for today was about the custom deleter you did for `std::unique_ptr` and exported it.
It is better to avoid exporting additional symbols if possible. And in this case, it is definitely possible.